Userland/useradd.cpp

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2019-2020, Jesse Buhagiar <jooster669@gmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <AK/String.h>
+#include <AK/StringBuilder.h>
+#include <LibCore/CArgsParser.h>
+#include <pwd.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+static constexpr uid_t BASE_UID = 1000;
+static constexpr gid_t USERS_GID = 100;
+const char* DEFAULT_SHELL = "/bin/sh";
+
+int main(int argc, char** argv)
+{
+    CArgsParser args_parser("useradd");
+    args_parser.add_arg("d", "home_dir", "home directory for the new user");
+    args_parser.add_arg("u", "uid", "user ID (uid) for the new user");
+    args_parser.add_arg("g", "group", "group ID (gid) for the new user");
+    args_parser.add_arg("m", "create home directory if it does not exist");
+    args_parser.add_arg("s", "shell", "name of default shell binary for the new user");
+    args_parser.add_arg("n", "name", "gecos name of the new user");
+    CArgsParserResult args = args_parser.parse(argc, argv);
+
+    if (argc <= 1) {
+        args_parser.print_usage();
+        return 0;
```

Comment:
The convention, at least the one I'm trying to follow, about outputting usage/help, is:

* When asked explicitly (i.e. with `--help`), output it to stdout and exit with 0
* When invoked in a wrong way, output it to stderr and exit with 1 (or other error code)

You don't have to change `CArgsParser::print_usage()` to support stdout vs stderr in this PR (if it doesn't already), but let's at least return non-0 here.

Reply:
> When asked explicitly (i.e. with --help), output it to stdout and exit with 0

Do we support `--` options? Or should I add `-help` to the args list?

Reply:
¯\\\_(ツ)\_/¯

I'm not asking you to implement `--help` support — that'd be nice, be IMO we can live without that, especially since there's a man page.

I'm saying, let's output usage to stderr (if possible) and exit with non-zero if usage wasn't explicitly asked for.

Reply:
> I'm saying, let's output usage to stderr (if possible) and exit with non-zero if we weren't explicitly asked for usage.

I'll get on that on another pr. I've changed the ret val to `1` instead of making it 0.